packages/stargate/src/queries/queryclient.ts

```diff
@@ -197,13 +197,18 @@ export class QueryClient {
     return value;
   }
 
-  public async queryRawProof(store: string, queryKey: Uint8Array): Promise<ProvenQuery> {
+  public async queryRawProof(
+    store: string,
+    queryKey: Uint8Array,
+    desiredHeight?: number,
+  ): Promise<ProvenQuery> {
     const { key, value, height, proof, code, log } = await this.tmClient.abciQuery({
       // we need the StoreKey for the module, not the module name
       // https://github.com/cosmos/cosmos-sdk/blob/8cab43c8120fec5200c3459cbf4a92017bb6f287/x/auth/types/keys.go#L12
       path: `/store/${store}/key`,
       data: queryKey,
       prove: true,
+      height: desiredHeight,
```

Comment:
How is it possible that state is available for past heights? Is every state key available for every height?

Reply:
It depends. You can keep every state (archive node) or only the current one (prunning=all).

Default is to keep historical state for the last 100 block I believe. They have a very complex database algorithm to maintain these historical state with proofs.

Reply:
Dig into iavl tree and look into orphan nodes and pruning